### ISSUE TITLE
Support using existing Docker images as hooks

### ISSUE THREAD

**Issue by jsok** [pre-commit/pre-commit#597]
Currently a `language: docker` hook only supports running a Docker image which it built itself (via a `Dockerfile`).

However this seems a little unnecessary and expensive since we can and should be building and publishing our Docker images to a registry.

Would there be interest in supporting using existing images as hooks? It should be simple and I'd be happy to make the PR.

## Proposed `.pre-commit-hooks.yaml` change

```yaml
-   id: flake8-docker
    name: flake8 (via docker)
    description: flake8, but provided by docker
    files: \.py$

    language: docker
    # Image to `docker pull`
    docker_image: my.registry.example.com/pre-commit-hooks/flake8:latest
    # Optional: use `ENTRYPOINT` of the image if not specified
    entry: flake8
```

**Comment by asottile** [comment]
First: thanks for the issue :) 

My main hesitation is I don't really want to add language specific configuration. 

I think you could simulate the same thing by having a trivial dockerfile with just `FROM <image>` which would largely get you the same result. Thoughts?

**Comment by jsok** [comment]
Yep totally get it 👍 

I think I can solve my underlying desire in a different way by just writing a wrapper script that calls `docker run` and users can override the image if desired using `args`.

**Comment by asottile** [comment]
@jsok actually now that I think about this more, I might be able to write a `language` which makes this work better -- would something like this be useful?

```yaml
-   id: flake8-docker
    name: flake8 (via docker)
    language: docker_image
    entry: my.registry.example.com/pre-commit-hooks/flake8:latest
    files: [python]
```

A downside would be that it would have to *depend* on the `ENTRYPOINT` of the image? so maybe not the best idea?

**Comment by jsok** [comment]
Ideally a Docker image and a git repo should be both considered valid plugin sources.
Creating a whole git repo just to act as a shim/wrapper for calling a Docker image is a bit tedious.

e.g. I'd love to see:

```yaml
-   repo: git://github.com/pre-commit/pre-commit-hooks
    sha: v0.9.1
    hooks:
    -   id: trailing-whitespace
-   docker: my.registry.example.com/pre-commit-images/flake8:python3
    hooks:
    -   id: flake8
```

---

As a back story, I use pre-commit as a linting step in my CI job rather than a git pre-commit hook so I'm probably not the typical use case. I've found pre-commit to be a simple and low-barrier entry for projects and team members to add linting steps.
Docker is great here because I don't need to install e.g. every version of python on my CI agents.
I want to let developers use **any** linter they like so long as they can package it as a docker image.

**Comment by asottile** [comment]
The `docker_image` source I'm proposing above melds nicely with [`local` hooks](http://pre-commit.com/#repository-local-hooks) as well which would obviate the need for a separate repository.  I guess my question is: would it satisfy your usecase well enough (before I put in some elbow grease to implement such a thing)?

Usage in CI is a pretty common situation too -- pre-commit is great at installing things :D (I usually do both CI + git hooks).

**Comment by jsok** [comment]
Yep I believe that would work well.

As for being locked into the one entrypoint, you could probably get around this limitation by building a docker image with no `ENTRYPOINT` and using the hooks `args` to specify the docker run `CMD`.

**Comment by asottile** [comment]
Oh yes of course!  I was trying to come up with a way to shimmy in `args: [--entrypoint, ...]` and didn't even think of the simple possibility of **not** having an `ENTRYPOINT`!  clever!

It should be pretty easy to implement, I'll see if I can't throw something together :)

**Comment by asottile** [comment]
Here's a PR which attempts to solve this :) https://github.com/pre-commit/pre-commit/pull/598

**Comment by asottile** [comment]
Documented here: https://github.com/pre-commit/pre-commit.github.io/pull/127

Available in 0.18.0

Let me know how it works out for you :)  Thanks again for the issue!

**Comment by jsok** [comment]
I've had a play with 0.18.0 and it seems to fit the bill 👍 
Appreciate the feature implementation and quick feedback.
